File: Week 02/id_011/binary-tree-preorder-traversal.go
package algorithm00401

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func preorderTraversal(root *TreeNode) []int {
	if root == nil {
		return nil
	}

	traversal := make([]int, 0)
	traversal = append(traversal, root.Val)
	traversal = append(traversal, preorderTraversal(root.Left)...)
	traversal = append(traversal, preorderTraversal(root.Right)...)

	return traversal
}

File: Week 02/id_161/BinaryTreePreorderTraversal.java
//给定一个二叉树，返回它的 前序 遍历。 
//
// 示例: 
//
// 输入: [1,null,2,3]  
//   1
//    \
//     2
//    /
//   3 
//
//输出: [1,2,3]
// 
//
// 进阶: 递归算法很简单，你可以通过迭代算法完成吗？ 
// Related Topics 栈 树
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> list = new LinkedList<>();
        traversal(root, list);
        return list;
    }
    public void traversal(TreeNode node, List<Integer> list) {
        if (node != null) {
            list.add(node.val);
            if (node.left != null) {
                traversal(node.left, list);
            }
            if (node.right != null) {
                traversal(node.right, list);
            }
        }
    }
}
